Pyxis kerberos authentication

FernandesMF opened the pull request:
## Pyxis kerberos authentication
Pyxis supports authentication through kerberos.  This method is preffered over SSL certificates because we won't need to renew our certificates periodically anymore.

This contribution implements kerberos authentication to the `pyxis-gql.py` module through the `requests-kerberos` library.

The Jira card associated with this contribution is https://issues.redhat.com/browse/CWFHEALTH-1762

### Testing remarks

I ran all the unit tests successfully. I also made manual e2e tests to make sure all queries were working properly. I used my personal kerberos credentials and send the requests to the dev pyxis gql endpoint.

I called all methods of the PyxisGQL class except for `query`, `_get_repo_projection` and `_get_image_projection`, which are called by the other methods. I ran the other methods just as in the unit tests and most of them returned empty lists as result (since the queried values were just mock values).

The methods `find_images_by_nvr` and `find_repositories` returned actual results. `find_repositories` took a pretty long time, and the result was quite long. I'm quoting an excerpt of the result from `find_repositories` below:
```python
[
 ...
 {'auto_rebuild_tags': [],
  'registry': 'registry.access.redhat.com',
  'release_categories': None,
  'repository': 'rhacm2/acm-cluster-templates-rhel8-operator'}
]  
```

### Questions/Technical debts
There are a few important things that I am leaving as technical debt for now:
 - requests-kerberos relies in a TGT that must be present in the execution environment; should we make this step in the Dockerfile, or is it already done?
 -  during the tests, I used my personal kerberos credentials; does freshmaker have a service credential, or should we request one?
 - the current implementation does not verify SSL certificates of the pyxis host; is this a requirement?
 -  the current implementation does not make mutual kerberos authentication; is this a requirement? (I pondered that, since we do not send sensitive data, it might not be needed, but it would be good to see other takes on this)

Please let me know if I should resolve any of these issues before we can merge!

qixiang commented on freshmaker/pyxis_gql.py:
```diff
@@ -47,10 +48,9 @@ def connect(self):
-            url=url,
-            retries=3,
-            cert=cert,
+            url=url, retries=3, cert=cert, auth=pyxis_krb_auth, verify=False
```
I don't remember why we have `verify=False`, we can remove this.

FernandesMF replied:
I believe this flag turns off the SSL verification during the requests. If I don't set it to false, I run into the following error when I try to make any of the queries:

> `requests.exceptions.SSLError: HTTPSConnectionPool(host='graphql.pyxis.dev.engineering.redhat.com', port=443): Max retries exceeded with url: /graphql/ (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:997)'))) `

I believe this flag will be necessary.

qixiang commented:
> requests-kerberos relies in a TGT that must be present in the execution environment; should we make this step in the Dockerfile, or is it already done?

We use a keytab to get the ticket and produce a credential cache (`conf.krb_auth_ccache_file`), it will be obtained automatically when authenticate with kerberos.

> during the tests, I used my personal kerberos credentials; does freshmaker have a service credential, or should we request one?

It already exists, so we don't need to worry about it.

> the current implementation does not verify SSL certificates of the pyxis host; is this a requirement?

No, it's probably a mistake while writing that code, it can be removed.

> the current implementation does not make mutual kerberos authentication; is this a requirement? (I pondered that, since we do not send sensitive data, it might not be needed, but it would be good to see other takes on this)

We have some APIs from Errata tool don't work with mutual kerberos authentication, so we just use `mutual_authentication=OPTIONAL` all around.

qixiang commented:
Beside of the comments, please also:
1. Squash some of the commits, we don't have auto squash enabled in this repo, so the developer is responsible for squash the commits in reasonable way. For example, the 6 commits in this PR could be squashed to 2 commits, one for the gitignore change, and the others should be squashed to a single commit.
2. Include the Jira key in commit message if the change is for a jira issue, to provide more info when others check the commit. Example:
```
A short summary

Meaningful descriptions

JIRA: CWFHEALTH-1762
```

FernandesMF commented on freshmaker/pyxis_gql.py:
```diff
@@ -47,10 +48,9 @@ def connect(self):
 class PyxisGQL:
     def __init__(self, url, cert):
```
Good point!

I removed the 'cert' parameter from PyxisGQL and cascaded this change to the rest of the code. I had to make a minor change to the ContainerAPI class in the container.py module, since it received a parameter that was just forwarded as 'cert'.

I also updated the tests for the container and pyxis_gql modules.

FernandesMF commented on freshmaker/pyxis_gql.py:
```diff
@@ -47,10 +48,9 @@ def connect(self):
     def __init__(self, url, cert):
         """Create authenticated Pyxis GraphQL session"""
+        pyxis_krb_auth = HTTPKerberosAuth(mutual_authentication=OPTIONAL, force_preemptive=True)
         transport = RequestsHTTPTransportWithCert(
```
Good point as well! I switched back to `RequestHTTPTransport as you suggested, thanks!

FernandesMF commented:
I addressed all the review points and squashed all my commits as Qixiang suggested.

I will open a new pull request from a new branch, though. I figured it would be easier to open a new pull request from a new branch than force-pushing these squashed changes to this branch.

FernandesMF commented:
Let's continue the review in #204
